src/relax/op/tensor/transform.cc

```diff
@@ -0,0 +1,441 @@
+    new_shape_prod *= inferred_len;
+    new_shape.Set(dim_to_infer, inferred_len);
+  }
+
+  if (analyzer->CanProve(old_shape_prod.value() != new_shape_prod)) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "Reshape expects the new shape to be convertible from the old shape. "
+                        "However, the old shape is "
+                     << data_sinfo->shape << ", with product " << old_shape_prod.value()
+                     << ", while the new shape is " << new_shape << ", with product "
+                     << new_shape_prod);
+  }
+
+  return TensorStructInfo(ShapeExpr(new_shape), data_sinfo->dtype);
+}
+
+TVM_REGISTER_OP("relax.reshape")
+    .set_num_inputs(2)
+    .add_argument("data", "Tensor", "The input tensor.")
+    .add_argument("new_shape", "Shape", "The input new shape.")
+    .set_attr<FInferStructInfo>("FInferStructInfo", InferStructInfoReshape);
+
+/* relax.transpose */
+TVM_REGISTER_NODE_TYPE(TransposeAttrs);
+
+Expr MakeTranspose(Expr data, Optional<Array<Integer>> axes) {
+  ObjectPtr<TransposeAttrs> attrs = make_object<TransposeAttrs>();
+  attrs->axes = std::move(axes);
+
+  static const Op& op = Op::Get("relax.transpose");
+  return Call(op, {std::move(data)}, Attrs{attrs}, {});
+}
+
+TVM_REGISTER_GLOBAL("relax.op.transpose").set_body_typed(MakeTranspose);
+
+bool IsIdentityPermutation(const std::vector<int>& permutation) {
+  for (int i = 0; i < static_cast<int>(permutation.size()); ++i) {
+    if (permutation[i] != i) {
+      return false;
+    }
+  }
+  return true;
+}
+
+StructInfo InferStructInfoTranspose(const Call& call, const BlockBuilder& ctx) {
+  TensorStructInfo data_sinfo = GetUnaryInputTensorStructInfo(call, ctx);
+
+  const auto* attrs = call->attrs.as<TransposeAttrs>();
+  if (data_sinfo->IsUnknownNdim()) {
+    if (attrs->axes.defined()) {
+      ctx->ReportFatal(Diagnostic::Error(call)
+                       << "Transpose cannot be performed when the input tensor " << data_sinfo
+                       << " ndim is unknown while the given number of axes " << attrs->axes.value()
+                       << " is clear. Please use MatchCast to match the input tensor to a specific "
+                          "ndim before doing transpose.");
+    }
+    return TensorStructInfo(data_sinfo->dtype, kUnknownNDim);
+  }
+
+  if (attrs->axes.defined()) {
+    int n_axis = attrs->axes.value().size();
+    if (n_axis != data_sinfo->ndim) {
+      ctx->ReportFatal(Diagnostic::Error(call)
+                       << "Transpose expects the number of input axes to equal the ndim of the "
+                          "input tensor. However, the tensor ndim is "
+                       << data_sinfo->ndim << " while the given number of axes is " << n_axis);
+    }
+  }
+
+  std::vector<int> axes;
+  if (attrs->axes.defined()) {
+    axes = CheckAxesInRangeNonRepetitive(call, ctx, data_sinfo->ndim, attrs->axes.value());
```

Comment:
Naming, perhaps rename to NormalizeAxes, since the main purpose is to normalize non-negative values. The checks are part of normalization process